Adding HNSW-QG AVX512f version

Patrick-H-Chen opened the pull request:
*Issue #, if available:*

Description of changes:

This change reflects the implementation of product quantization with AVX512f SIMD on top of HNSW to perform fast quantization-based graph search. Following are two major changes:

1) Adding a ProductQuantizer with 4 bits local centroids.
2) A HNSWProductQuantizer4Bits struct which leverages ProductQuantization as the distance calculation instead of original full L2 distance.
 

 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

OctoberChang commented on pecos/core/ann/hnsw.hpp:
```diff
@@ -238,6 +243,142 @@ namespace ann {
         }
     };
 
+    template<class FeatVec_T>
```
Do we still need the template syntax, if we are restricting to dense vector only?

Patrick-H-Chen replied:
We will firstly restrict the implementation to FeatVecDenseL2Simd<float>, and extend it later when other metric is implemented.

OctoberChang commented on pecos/core/ann/hnsw.hpp:
```diff
@@ -870,5 +1009,343 @@ namespace ann {
+            nlohmann::json j_params = {
+                {"hnsw_t", pecos::type_util::full_name<HNSWProductQuantizer4Bits>()},
+                {"version", "v1.0"},
+                {"train_params", {
```
should we also save other arguments passed into the train interface (e.g., `subspace_dimension`)?

Patrick-H-Chen replied:
Resolved in next commit.

OctoberChang commented on pecos/core/ann/quantizer_impl/common.hpp:
```diff
@@ -0,0 +1,237 @@
+
+                // fit HLT or flat-Kmeans for each sub-space
+                std::vector<index_type> assignments(sub_sample_points);
+                pecos::clustering::Tree hlt(4);
```
avoid hard-coded, try `log2(num_of_local_centroids)`.

Patrick-H-Chen replied:
Resolved in next commit.

OctoberChang commented on pecos/core/ann/hnsw.hpp:
```diff
@@ -870,5 +1009,343 @@ namespace ann {
-} // end of namespace pecos
+    template<typename dist_t>
+    struct HNSWProductQuantizer4Bits {
+        typedef FeatVecDenseL2Simd<float> feat_vec_t;
```
we currently only support `L2`, need to add some warning.

Patrick-H-Chen replied:
Currently we don't have parameter to let user input distance metric. Maybe it's easier to write it in the README.md in the example folder ?

OctoberChang commented on pecos/core/ann/hnsw.hpp:
```diff
@@ -870,5 +1009,343 @@ namespace ann {
+            memcpy(graph_l1.buffer.data(), hnsw.graph_l1.buffer.data(), hnsw.graph_l1.buffer.size() * sizeof(index_type));
+
+            graph_l0_pq4.build_quantizer(X_trn, subspace_dimension, sub_sample_points);
+            graph_l0_pq4.build_graph(hnsw.graph_l0);
```
should we `delete hnsw` to reduce data from 3x to 2x?

OctoberChang commented on pecos/core/ann/quantizer_impl/x86.hpp:
```diff
@@ -0,0 +1,227 @@
+        }
+
+        __attribute__((__target__("avx512f")))
+        inline void setup_lut(float* query, uint8_t* lut_ptr, float& scale, float& bias) const {
```
consider renaming `lut_ptr` to `lut` (as in Line 154) or vice versa for naming consistency.

Patrick-H-Chen replied:
lut refers to std::vector<float> and lut_ptr is lut.data(). If I don't use it as a function argument, anyway I need to initiate a pointer to traverse the lut.

OctoberChang commented on pecos/core/ann/quantizer_impl/x86.hpp:
```diff
@@ -0,0 +1,227 @@
+            pad_parameters_default(max_degree, code_dimension);
+        }
+
+        __attribute__((__target__("avx512f"))) void pad_parameters(index_type& max_degree, index_type& code_dimension) {
```
consider a newline before `void pad_parameters()`

OctoberChang commented on pecos/core/ann/quantizer_impl/x86.hpp:
```diff
@@ -0,0 +1,227 @@
+
+        __attribute__((__target__("avx512f"))) void pad_parameters(index_type& max_degree, index_type& code_dimension) {
+            //  When using AVX512f, we have 16 centroids per local codebook, and each of it uses 8 bits to represent quantized
+            //  distance value. Thus,m we will have 128 bits to load 1 set of local codebooks. Thus, a loadu_si512 will load
```
typo: `Thus,m` -> `Thus,`

OctoberChang commented on pecos/core/ann/quantizer_impl/x86.hpp:
```diff
@@ -0,0 +1,227 @@
+        }
+
+        __attribute__((__target__("default")))
+        inline void approximate_neighbor_group_distance(size_t neighbor_size, float* ds, const char* neighbor_codes, uint8_t* lut_ptr, float scale, float bias) const {
```
naming consistency: `lut_ptr` or `lut`?

Patrick-H-Chen replied:
resolve in the next commit.

OctoberChang commented on pecos/core/ann/quantizer_impl/x86.hpp:
```diff
@@ -0,0 +1,227 @@
+        __attribute__((__target__("avx512f")))
+        inline void approximate_neighbor_group_distance(size_t neighbor_size, float* ds, const char* neighbor_codes, uint8_t* lut_ptr, float scale, float bias) const {
+            //  When using AVX512f, we have 16 centroids per local codebook, and each of it uses 8 bits to represent quantized
+            //  distance value. Thus,m we will have 128 bits to load 1 set of local codebooks. Thus, a loadu_si512 will load
```
typo: `Thus,m` -> `Thus,`

OctoberChang commented on pecos/core/ann/quantizer_impl/x86.hpp:
```diff
@@ -0,0 +1,227 @@
+            // Similarly, we have to parse every 16 neighbors at a time to maximally leverage avx512f.
+            // Thus we can also calculate the number of group iterations based on if neighbor_size id
+            // divisible by 16.
+            index_type num_groups = neighbor_size % 16 == 0 ? neighbor_size / 16 : neighbor_size / 16 + 1;
```
This seems to be handled during Line 86 (the `pad_parameters()` function) ? If so, we can just do `num_groups=neighbor_size % 16`?

Patrick-H-Chen replied:
num_groups depends on neighbor_size, and Line 86 consider the maximal neighbors (max_degree). So it's possible max neighbors is 96, but for a node it only has say 18 neighbors, which is not divisible by 16. So we still need to check again here.

OctoberChang commented on pecos/core/ann/quantizer_impl/common.hpp:
```diff
@@ -0,0 +1,237 @@
+            if (dimension % M != 0) {
+                throw std::runtime_error("Original dimension must be divided by subspace dimension");
+            }
+            num_local_codebooks = M;
```
consider just replace `M` by `num_local_codebooks`, and use that naming consistently in other place.

Patrick-H-Chen replied:
Resolve in the next commit.

OctoberChang commented on pecos/core/ann/quantizer_impl/x86.hpp:
```diff
@@ -0,0 +1,227 @@
+            //  distance value. Thus,m we will have 128 bits to load 1 set of local codebooks. Thus, a loadu_si512 will load
+            //  512 / 128 == 4 local codebooks at a time. Thus, will need to be adjust the code_dimension to make it divisible
+            //  by 4. If it's not, we have to pad 0 to extended dimensions.
+            code_dimension = code_dimension % 4 == 0 ? code_dimension : (code_dimension / 4 + 1) * 4;
```
what's the relation between `code_dimension` versus `local_dimension` ?

Patrick-H-Chen replied:
Same as below, code_dimension is from the perspective of HNSWPQ4Bits and local_dimension is from the perspective of quantizer. I will not pad  local_dimension but code_dimension could be padded if to fit memory layout in GraphPQ4Bits. In sum, code_dimension could be different from local_dimension.

OctoberChang commented on pecos/core/ann/hnsw.hpp:
```diff
@@ -238,6 +243,143 @@ namespace ann {
+            }
+            // currently, we don't support padding 0 on X_trn, so the cols of X_trn must be divisible by subspace_dimension.
+            // otherwise, we will throw error in quantizer.train().
+            quantizer.train(X_trn, code_dimension, sub_sample_points);
```
in this case, `code_dimension` is actually `num_local_codebook` ?!

OctoberChang replied:
consider naming variables for consistency in `quantizer.hpp`

Patrick-H-Chen replied:
code_dimension is from the perspective of HNSWPQ4Bits and num_local_codebook is from the perspective of Quantizer. So code_dimension could be padded to fit memory layout in GraphHNSWPQ4 but num_local_codebook will not be padded.

Patrick-H-Chen replied:
After reading again the code, I think renaming code_dimension into num_local_codebook_pad is actually not a good idea. From Graph construction perspective, it's the number of dimensions a code will use, and there is no concept of "local codebook" from Graph's view. In addition, when calculating the memory space for neighboring code, code_dimension is a much more meaningful name than num_local_codebook_pad. So I think the name could be kept same.

OctoberChang replied:
I see. In such case, I would suggest you write some comments that describe the definitions of those variables under those structures (e.g., GraphL0PQ4bit, HNSWProdQuad4Bit, quantizer)

Patrick-H-Chen replied:
Revise in the next commit.

OctoberChang commented on pecos/core/ann/hnsw.hpp:
```diff
@@ -238,6 +243,143 @@ namespace ann {
+
+        ProductQuantizer4Bits quantizer;
+        index_type num_node;
+        index_type code_dimension;
```
is this equivalent to `local_dimension` as defined `quantizer struct`?

Patrick-H-Chen replied:
local_dimension is the number of dimensions of subspace, but code_dimension is the code length encoded in the graph for a single data point. code_dimension is the padded num_local_codebooks, and as explained above I think it's better not to rename.

rootlesstree committed:
Initial fix of attribute ‘target(default)’ is not valid

2nd fix of attribute ‘target(default)’ is not valid

3rd fix by moving pad_parameters into quantizer

Revise Quantizer by adding commoh.hpp

fix typo in default quantizer

Revise FeatVec_T and Quantizer #centroids

revise quantizer train, and some other minor fixes

add some comments on variables

Adding HNSW-QG AVX512f version

add type checker
